gossip/src/crds_value.rs

```diff
@@ -490,84 +493,164 @@ impl Sanitize for NodeInstance {
+                if *current_bytes > RestartLastVotedForkSlots::MAX_BYTES as u32 {
+                    return None;
+                }
+                Some(U16(count))
+            })
+            .collect();
+        Self { encoded }
+    }
+
+    pub fn slots_count(&self) -> usize {
+        self.encoded.iter().map(|x| usize::from(x.0)).sum::<usize>()
+    }
+
+    pub fn to_slots(&self, last_slot: Slot, min_slot: Slot) -> Vec<Slot> {
+        let mut result: Vec<Slot> = self
+            .encoded
+            .iter()
+            .zip([1, 0].iter().cycle())
+            .flat_map(|(bit_count, bit)| std::iter::repeat(bit).take(usize::from(bit_count.0)))
+            .enumerate()
+            .filter(|(_, bit)| **bit > 0)
+            .map(|(offset, _)| last_slot.saturating_sub(offset as Slot))
+            .take(MAX_SLOTS_PER_ENTRY)
+            .take_while(|slot| *slot >= min_slot)
+            .collect();
+        result.reverse();
+        result
+    }
+}
+
+#[derive(Deserialize, Serialize, Clone, Debug, PartialEq, Eq, AbiExample)]
+pub struct RawOffsets {
+    bits: BitVec<u8>,
+}
+
+impl RawOffsets {
+    pub fn new(mut bits: BitVec<u8>) -> Self {
+        bits.truncate(RestartLastVotedForkSlots::MAX_BYTES as u64 * 8);
+        bits.shrink_to_fit();
+        Self { bits }
+    }
+
+    pub fn to_slots(&self, last_slot: Slot, min_slot: Slot) -> Vec<Slot> {
+        let mut result: Vec<Slot> = (0..self.bits.len())
+            .filter(|index| self.bits.get(*index))
+            .map(|offset| last_slot.saturating_sub(offset))
+            .take_while(|slot| *slot >= min_slot)
+            .take(MAX_SLOTS_PER_ENTRY)
+            .collect();
+        result.reverse();
+        result
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone, PartialEq, Eq, AbiExample, Debug)]
 pub struct RestartLastVotedForkSlots {
     pub from: Pubkey,
     pub wallclock: u64,
-    pub slots: Vec<CompressedSlots>,
+    offsets: SlotsOffsets,
+    pub last_voted_slot: Slot,
     pub last_voted_hash: Hash,
     pub shred_version: u16,
 }
 
 impl Sanitize for RestartLastVotedForkSlots {
     fn sanitize(&self) -> std::result::Result<(), SanitizeError> {
-        if self.slots.is_empty() {
-            return Err(SanitizeError::InvalidValue);
-        }
-        self.slots.sanitize()?;
         self.last_voted_hash.sanitize()
     }
 }
 
 impl RestartLastVotedForkSlots {
-    pub fn new(from: Pubkey, now: u64, last_voted_hash: Hash, shred_version: u16) -> Self {
-        Self {
+    // This number is MAX_CRDS_OBJECT_SIZE - empty serialized RestartLastVotedForkSlots.
+    const MAX_BYTES: usize = 824;
+
+    pub fn new(
+        from: Pubkey,
+        now: u64,
+        last_voted_fork: &[Slot],
+        last_voted_hash: Hash,
+        shred_version: u16,
+    ) -> Result<Self, String> {
+        let Some((first_voted_slot, last_voted_slot)) =
+            last_voted_fork.iter().minmax().into_option()
```

Comment:
`.copied()` at the end, so that `(first_voted_slot, last_voted_slot)` are not reference to POD.
https://doc.rust-lang.org/std/option/enum.Option.html#method.copied-1

Reply:
Hmm, that doesn't work:
   --> gossip/src/crds_value.rs:602:59
    |
602 |             last_voted_fork.iter().minmax().into_option().copied()
    |                                                           ^^^^^^ method cannot be called on `Option<(&u64, &u64)>` due to unsatisfied trait bounds
    |
   ::: /Users/wen/.rustup/toolchains/1.73.0-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/option.rs:563:1
    |
563 | pub enum Option<T> {
    | ------------------ doesn't satisfy `std::option::Option<(&u64, &u64)>: Iterator`
    |
    = note: the following trait bounds were not satisfied:
            `std::option::Option<(&u64, &u64)>: Iterator`
            which is required by `&mut std::option::Option<(&u64, &u64)>: Iterator`

Reply:
Then I think you can do something like this:
```rust
let Some((&first_voted_slot, &last_voted_slot)) =
    last_voted_fork.iter().minmax().into_option()
else {
  // ...
};
```

Reply:
Done.